NovaCustom MiniPC

filipleple opened the pull request.

filipleple committed:
mainboard/novacustom/nuc_box/: add

This is a copy of the System76 Meer9 code, which is an initial port
of the ASROCK NUC BOX hardware. This platform is also based on that
hardware.

Upstream-Status: Inappropriate [Dasharo downstream]
Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

mkopec commented:
If we're gonna copy code from somewhere, please give proper attribution at least in the commit message

mkopec commented on src/mainboard/novacustom/nuc_box/dsdt.asl:
```diff
@@ -0,0 +1,36 @@
+/* SPDX-License-Identifier: GPL-2.0-only */
+
+//TODO: HACK FOR MISSING MISCCFG_GPIO_PM_CONFIG_BITS
```
should not be needed here

mkopec commented on src/mainboard/novacustom/nuc_box/acpi/s76.asl:
this entire file should be removed

filipleple committed:
configs/config.novacustom_nuc_box: extend for feature parity

The NVC NUC BOX is planned to provide maximum possible feature parity
with other NovaCustom platforms, especially the Meteor Lake laptops.
This commit enables most features in the platform config.

Upstream-Status: Inappropriate [Dasharo downstream]
Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

filipleple commented:
@mkopec https://github.com/Dasharo/coreboot/pull/643/commits/eafd734cfbbf672273c646e6f31ef68a4bfcf706

filipleple committed:
build.sh: add NovaCustom NUC BOX

Adds the NUC BOX to build.sh, for convenient automatic building via the
command ./build.sh nuc_box

Upstream-Status: Inappropriate [Dasharo downstream]
Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

filipleple committed:
mainboard/novacustom/nuc_box/bootblock.c: implement fan curve options

Introduce support for multiple fan curve profiles via
CONFIG_EDK2_FAN_CURVE_OPTION. Fan behavior can now be adjusted at
bootblock stage using one of the following modes:

Silent (default) – conservative ramp-up, prioritizes quiet operation
Performance – aggressive cooling response
Off – disables active cooling unless critical temperature is reached

This is implemented via PECI-based Smart Fan IV configuration in the
hardware monitor setup.

Upstream-Status: Inappropriate [Dasharo downstream]
Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

filipleple committed:
mainboard/novacustom/nuc_box/variants/nuc_box/board.fmd: add for BIOS lock

This file is required for vboot, as well as proper BIOS lock and SMM BWP
functionality. It includes the console region, since the board doesn't
have any debug port, making the SPI flash the only option for debug
logging during development.

Upstream-Status: Inappropriate [Dasharo downstream]
Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

filipleple committed:
mainboard/novacustom/nuc_box/Kconfig: set vboot-related options

This commit adds some flags which are necessary for vboot to function
properly on a given board.

Upstream-Status: Inappropriate [Dasharo downstream]
Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

miczyg1 commented on src/mainboard/novacustom/nuc_box/ramstage.c:
```diff
@@ -0,0 +1,18 @@
+
+static void mainboard_init(void *chip_info)
+{
+	mainboard_configure_gpios();
```
This is too late. At this point the FSP will be long done. If we program GPIOS that late, FSP may fail to initialize certain IP blocks. It should happen before FSP-M ideally.

filipleple replied:
https://github.com/Dasharo/coreboot/pull/643/commits/ae57fc982cdcb6b719cab509a6a949ed4204742f

miczyg1 commented on src/mainboard/novacustom/nuc_box/variants/nuc_box/ramstage.c:
```diff
@@ -0,0 +1,18 @@
+
+	// XXX: Enabling C10 reporting causes system to constantly enter and
+	// exit opportunistic suspend when idle.
+	params->PchEspiHostC10ReportEnable = 0;
```
if D3COLD_SUPPORT is set to disabled, then no opportunistic suspend will happen. It seems to be broken board code all along...

This option and D3COLD_SUPPORT should be enabled for opportunistic suspend to work...

filipleple replied:
https://github.com/Dasharo/coreboot/pull/643/commits/ae57fc982cdcb6b719cab509a6a949ed4204742f

miczyg1 commented on src/mainboard/novacustom/nuc_box/variants/nuc_box/romstage.c:
```diff
@@ -0,0 +1,25 @@
+	const bool half_populated = false;
+
+	mupd->FspmConfig.DmiMaxLinkSpeed = 4;
+	mupd->FspmConfig.GpioOverride = 0;
```
This should not be cleared. By default coreboot ste's it to 1, so that FSP will not override coreboot programmed GPIOs.

Remove it and initialize GPIOs before FSP-M. Then some problems will possibly go away.

filipleple replied:
https://github.com/Dasharo/coreboot/pull/643/commits/ae57fc982cdcb6b719cab509a6a949ed4204742f

miczyg1 commented on 3rdparty/dasharo-blobs:
Why not reuse FSP and GOP driver from the MTL laptops?

filipleple replied:
Will try with the FSP, the laptop GOP driver doesn't work - it somehow alters the VBT (I've compared what ends up loaded by Ubuntu to what's in the repo) making the HDMI ports not work.

miczyg1 replied:
VBT is always altered when seen by OS. VBT is a part of operation region that is used for communicating with graphics. You can't expect to have the same VBT after intiialization, as it is supposed to be patched runtime.

Also I don;t see a reason to use EFI GOP driver at all. FSP GOP should be sufficient in the first place. We used EFI GOP in laptops because of high DPI panel. The NUC BOX does not have internal LCD panel

filipleple replied:
Alright, I've tried the laptop FSP and as expected it did work just fine. However, without the GOP (removing the appropriate `NO_GFX_INIT` and `GOP_FILE` flags from defconfig), we get a black screen with no new errors in dmesg/cbmem.

miczyg1 replied:
> Alright, I've tried the laptop FSP and as expected it did work just fine. However, without the GOP (removing the appropriate `NO_GFX_INIT` and `GOP_FILE` flags from defconfig), we get a black screen with no new errors in dmesg/cbmem.

Removing some options from defconfig doesn't mean the build is configured properly. Defconfig shouldn't be modified manually. If you need to change something you do:

```
cp configs/config.<platform> .config
make menuconfig
<modify settings now and save>
make olddefconfig
cp defconfig configs/config.<platform>
```

filipleple replied:
Right, this is definitely worth knowing.

Still though, going the menuconfig route, doesn't work without the GOP

miczyg1 replied:
FSP needs proper VBT to get it working. Any mismatch in versions may lead to lack of display, which I recently encountered on ODROID. Maybe here is the same problem.

filipleple replied:
https://github.com/Dasharo/dasharo-issues/issues/1417

filipleple committed:
mainboard/novacustom/nuc_box/: initialize GPIOs before FSP-M

If we program the GPIOs too late, FSP may fail to initialize certain IP blocks.
The programming should happen before FSP-M.

Upstream-Status: Inappropriate [Dasharo downstream]
Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

filipleple commented:
So the platform is operational, but only with vboot disabled. I thought to separate the effort we could review and merge in this state, then I could work on vboot integration in a consecutive PR. Moreover, we might even end up resigning from using vboot on this platform considering the new firmware redundancy feature plan.

miczyg1 commented on src/mainboard/novacustom/nuc_box/Kconfig:
```diff
@@ -0,0 +1,102 @@
+	select VBOOT_CLEAR_RECOVERY_IN_RAMSTAGE
+	select VBOOT_MOCK_SECDATA
+	select VBOOT_NO_BOARD_SUPPORT
+	select VBOOT_SLOTS_RW_A
```
`VBOOT_SLOTS_RW_A` should be allowed to be selectable from menu.

filipleple replied:
done https://github.com/Dasharo/coreboot/pull/643/commits/d10f1fc755bd6846d183be51f4b791dd738f6b12

krystian-hebel commented on build.sh:
```diff
@@ -452,7 +453,11 @@ case "$CMD" in
+        BOARD="nuc_box"
+        build_novacustom "nuc_box"
+        ;;
+    "v540tu " | "V540TU " )
```
```suggestion
    "v540tu" | "V540TU" )
```

filipleple replied:
https://github.com/Dasharo/coreboot/pull/643/commits/cb4ff64b934b0c02e7944bc304f6444bb2b2f170

filipleple committed:
3rdparty/dasharo-blobs: bump for NUC BOX blobs

Provides access to the blobs required by a NUC BOX build,
namely the IFD, ME and GOP

Upstream-Status: Inappropriate [Dasharo downstream]
Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

mkopec committed:
configs: Update to fix check-defconfigs error

Upstream-Status: Inappropriate [Dasharo downstream]
Change-Id: I3128d53ddb7e9a5f8095df5fc66c58e1412cc48f
Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>
